core/modules/replica/SqlEnableDbJob.cc

```diff
@@ -0,0 +1,143 @@
+ *
+ * You should have received a copy of the LSST License Statement and
+ * the GNU General Public License along with this program.  If not,
+ * see <http://www.lsstcorp.org/LegalNotices/>.
+ */
+
+// Class header
+#include "replica/SqlEnableDbJob.h"
+
+// System headers
+#include <algorithm>
+#include <stdexcept>
+
+// Qserv headers
+#include "replica/Configuration.h"
+#include "replica/ServiceProvider.h"
+#include "replica/SqlEnableDbRequest.h"
+#include "replica/StopRequest.h"
+
+// LSST headers
+#include "lsst/log/Log.h"
+
+using namespace std;
+
+namespace {
+
+LOG_LOGGER _log = LOG_GET("lsst.qserv.replica.SqlEnableDbJob");
+
+} /// namespace
+
+namespace lsst {
+namespace qserv {
+namespace replica {
+
+string SqlEnableDbJob::typeName() { return "SqlEnableDbJob"; }
+
+
+SqlEnableDbJob::Ptr SqlEnableDbJob::create(
+        string const& database,
+        bool allWorkers,
+        Controller::Ptr const& controller,
+        string const& parentJobId,
+        CallbackType const& onFinish,
+        Job::Options const& options) {
+
+    return Ptr(new SqlEnableDbJob(
+        database,
+        allWorkers,
+        controller,
+        parentJobId,
+        onFinish,
+        options
+    ));
+}
+
+
+SqlEnableDbJob::SqlEnableDbJob(string const& database,
+                               bool allWorkers,
+                               Controller::Ptr const& controller,
+                               string const& parentJobId,
+                               CallbackType const& onFinish,
+                               Job::Options const& options)
+    :   SqlJob(0,
+               allWorkers,
+               controller,
+               parentJobId,
+               "SQL_ENABLE_DATABASE",
+               options),
+        _database(database),
+        _onFinish(onFinish) {
+}
+
+
+list<pair<string,string>> SqlEnableDbJob::extendedPersistentState() const {
+    list<pair<string,string>> result;
+    result.emplace_back("database", database());
+    result.emplace_back("all_workers", string(allWorkers() ? "1" : "0"));
+    return result;
+}
+
+
+list<SqlRequest::Ptr> SqlEnableDbJob::launchRequests(util::Lock const& lock,
+                                                     string const& worker,
+                                                     size_t maxRequests) {
+
+    // Launch exactly one request per worker unless it was already
+    // launched earlier
+
+    list<SqlRequest::Ptr> requests;
+    if (not _workers.count(worker) and maxRequests != 0) {
+        auto const self = shared_from_base<SqlEnableDbJob>();
+        requests.push_back(
+            controller()->sqlEnableDb(
+                worker,
+                database(),
+                [self] (SqlEnableDbRequest::Ptr const& request) {
+                    self->onRequestFinish(request);
+                },
+                options(lock).priority,
+                true,   /* keepTracking*/
+                id()    /* jobId */
+            )
```

Comment:
It seems like it shouldn't be necessary for Controller to have a distinct function for every Job sub-class, but I'm not that familiar with the underlying code.

Reply:
`jobs` in the Replication System represent high-level algorithm which initiate operations on multiple (or all) workers of a set up. Jobs may initiate other jobs, or so called `requests` (each request is sent to one worker). Requests are instantiated and managed by class `Controller`. Each request has a type and it requires a specific set of input parameters. That is why there is a dedicated method in the `Controller`, so that it would be easier to search for all uses of these methods in the code ... if needed. Besides, the names of the methods suggest what requests are launched by them.
Note that this ticket is not about changing the API of the Controller, Requests or Jobs, it's about adding more Jobs to the Replication system module to support operations on behalf of the Ingest system.
More details on a design of the Replication System's framework could be found at:
https://confluence.lsstcorp.org/display/DM/Replication+System+design